### ISSUE TITLE
Config Overhaul 

### ISSUE THREAD

**Issue by mjmadsen** [PokemonGoF/PokemonGo-Bot#5218]
I've been giving a lot of thought to how we setup our configuration files. Here on my thoughts on how we can clean up a whole lot of what we have going on there, as well as our documentation. 

Our base config includes global goodies, such as our `debug`, `distance_unit`, `reconnecting_timeout`, etc. Could be split into like types (sleep, output, etc). Also includes a list of all tasks with a simple enable/disable option. 

A config file per task. When we update a task, we update a single doc about that task's configuration. If we have an update to the file, it'll be easy to realize where you need to make changes as it's not a mile long. This creates a whole lot more files, but they'll be less overwhelming. We could also add a bit to see if two enabled tasks conflict with each other (not quite on my point here). 

Next bit I'm open on, sleep timers for different tasks (possibly speeds too?) all in a different file separate from our task configuration. We change sleep ranges less often than anything else. Plus many of them are repeats, currently. For example, repeat same wait for a similar action in another task. 

Last, we could add some bits to limit our need for a hundred ways to filter pokemon and combine release, ignore, snipe, walk_to, transfer, upgrade and so on. Simplest change is just move all pokemon related settings into one file, merging release, transfer, and so forth are all under the same "pokemon" index. This would make for a very long file, but this file would hardly change unless we have some major refactoring. I have more ramblings out this, but I'm drifting...

Ex:
`      "Rattata": {
           "release_below_iv": 0.9, "release_below_cp": 200,
           "catch_above_cp": 0.5, "catch_above_iv": 0.0,
           "keep_best_cp": 1, "keep_best_iv": 1,
           "always_catch" : false,
           "always_release" : false,
            "evolve_above_cp": 500, "evolve_above_iv": 0.8,
},`

That's just a sample of what could be, I'm sure that would need tweaking and there are a billion other ways to go about this. Related to this, it would be nice if we could also set globals for these values. Say we didn't want to specify per pokemon keep_best values, but instead always keep the top cp/iv of all and also, keep the values listed for the pokemon. If that makes sense...

I'll stop.

**Comment by avexus** [comment]
I also noticed that
{"type": "CollectLevelUpReward",	"config": {"collect_reward": true,	"level_limit": -1}},
This is the example that other tasks all use "enable", and this uses "collect_reward". It's not a big issue, but change it just to match other tasks, and to look better.

**Comment by mjmadsen** [comment]
Nice catch :+1: 

@solderzzc @rawgni @Gobberwart @Anakin5 @th3w4y @askovpen @sohje @cmezh Your thoughts? I'd like a bit of input before I delve too deep in the makeover.

**Comment by crvfts** [comment]
A separate pokemon config would be VERY useful but I think it should have a few global variables and that those global variables have the ability to be enable: true/false. Sometimes I want to catch anything and everything, other times I only want to scour for very rares/ultra rares and I generally never want to combine those two efforts.

**Comment by sohje** [comment]
Great ideas! Really appreciate it! 👍 
Yeap but, we need to think about usability it become much easier to update confs/md/etc but if we split config into chunks, what about if _for example_ i want to run some bots using 1 conf file with release everything, another bots to release above CP/IV/etc, some bots will go to camp near the lure, another bots is about to snip everything they get from social/map/etc - what about this? We have to copy-paste a lot of files and change confs? Now i can make 1-N different confs and just run my bots. If we split every task into separate configs, we need to edit a lot of files. Maybe its not common usage behaviour, just thoughts ;)
p.s. sorry for typos, didnt sleep for long.

**Comment by mjmadsen** [comment]
Thanks for the input @sohje - that's a good points. Perhaps in the base config, we have a setting on which setup of the task you'd like. Then in task config we offer a few base samples. Say for releasing pokemon, we have a strict (keep only very best), general (keep most), and loose (keep all within range of bag size). Obviously those names could use some work though :smile: 

I suppose that could be expanded to allow users to create custom names - so our devs can setup lots of different variants of the task setup. Then in turn, our regular users who probably don't tweak stuff too much can choose a name that sounds appropriate for them (and they don't go messing with lots of settings).

**Comment by sohje** [comment]
I think for the first step of refactoring/splittin we can make some sort of _globals.json_ 🐙 
With default/safe? values.
```js
{
    "websocket_server": false,
    "heartbeat_threshold": 10,
    "enable_social": false,
    "live_config_update": {
      "enabled": true,
      "tasks_only": true
    },
    "walk_max": 11.11,
    "walk_min": 4.18,
    "alt_min": 500,
    "alt_max": 1000,
    "replicate_gps_xy_noise": false,
    "replicate_gps_z_noise": false,
    "gps_xy_noise_range": 0.000125,
    "gps_z_noise_range": 12.5,
    "health_record": false,
    "location_cache": true,
    "distance_unit": "m",
    "reconnecting_timeout": 15,
    "logging": {
      "color": true,
      "show_datetime": true,
      "show_process_name": true,
      "show_log_level": true,
      "show_thread_name": false
    },
    "spin_fort_delay_min": 2,
    "spin_fort_delay_max": 7,
    ....
}
```
And something like default config.json, if task name in dict object then its already enabled.

```js
{
    "HandleSoftBan": {},
    "LevelUpRewar": {},
    "CompleteTutorial": { "min_interval": 120 },
    "IncubateEggs": {},
    "UpdateLive": {
        'inventory': {},
        'stats': {}
    },
    "Release": { "option": "strict"},
    "LevelUpReward": {},
   ....
}

```

**Comment by Gobberwart** [comment]
I can see users ending up with tens (if not hundreds) of fragmented config files for different things, and I'm not sure this is really an improvement over monolithic configs. Just thinking through the various scenarios at the moment.

**Comment by Gobberwart** [comment]
Perhaps we could allow configs to be cumulative, eg. we have config.default.json (or a set of files) which are values we think are safe to use.

User could then specify addition files containing config information to be added to the defaults. 

Eg. ./pokecli.py --cf configs/config.default.json -cf configs/config.snipe.json -cf configs/config.runaroundincircles.json

Each file could be as small as one setting or a complete overwrite. 

Settings to be applied in order read from the command line.

**Comment by mjmadsen** [comment]
@Gobberwart your first comment helped me come up with an idea. We could have a script that runs when we crash, building a crash log. It includes commit, os data, crash log, time visited last pokestop, time since last pokemon encounter, so on, and based on crash log pulls related task settings. Then users could simply post that in their issue. This, however, wouldn't help for more subtle problem, like our polylinewalker moving too fast. Perhaps we could allow input, where a user could specify their issue (speed, release, transfer, evolve, whatever) and we build a log in the same fashion. IDK, in the real of possibility? Perhaps we could use this with our health_record - speaking of which, do we use this information for anything at the moment?

**Comment by Gobberwart** [comment]
@mjmadsen Good idea mate. Could we make use of this (or something like it): https://github.com/lobocv/crashreporter

**Comment by mjmadsen** [comment]
@Gobberwart I'll have to take a look at that. I was thinking we could diff our configs with that too, only list changes.

Perhaps things that are working well (sleep timers, action delays, etc) could be moved to and "advanced" directory. Same with other "maybe you shouldn't mess with this if you don't know what you're doing" related settings.

**Comment by rawgni** [comment]
most if not all of the workers should already have a default config inside their respective code. Thus, the default config should only show "enable/disable". If they were to tinker with the config, they can use the "advanced" directory.

**Comment by mjmadsen** [comment]
Ah, so our examples exclude certain settings? Those mostly being values most people won't change - like the default action time. But in our documentation, we list the potential settings (advanced users will read)

**Comment by MZorzy** [comment]
How about a aubdir per bot ?
ls /config/%botname%/
 config
/config/%botnam%/advanced
 catcht
 relase
 evolve
moving

So we use 
Start.sh my4bot
Or can mix config in command line like
Start.sh bot4 -evolve bot4

Other subdir like /config/auth
Ad config/location <-gps data with altitude range inside

**Comment by Gobberwart** [comment]
so...
config.default.json = basic common settings that should work for everyone at a simple level
advanced/config.whatever.json = configs for different tasks/options etc. that could be cumulatively added by "advanced" users

**Comment by Gobberwart** [comment]
@mzorzy subdir per bot could get complicated to maintain. Then again, if you are capable of running multiple bots with multiple configs, you should be able to cope :)

**Comment by MZorzy** [comment]
@gobberwart now only 4 bot, rhe biggest issue cp Nx config file due different gps altitude on different location. So 5 location 20 config files. Multiply if want other moving or relase/catch ...

**Comment by mjmadsen** [comment]
If we allow task setting config branches, we could have 2-3 default settings, and our advanced users could add more. This reduces a lot of redundancy if you're running a lot of bots with different settings. 
`"task_name": enabled,
"settings": "generic"
"some_setting": "override value from file"`

**Comment by Gobberwart** [comment]
Exactly. Bot can load multiple config files (passed from command line) and process them in the order received.

Eg. python ./pokecli.py -cf configs/defaults.json -cf configs/sniping.json -cf configs/evolveall.json

Loads defaults from defaults.json
Adds/overwrites settings from sniping.json
Adds/overwrites settings from evolveall.json
-> Final cumulative config

**Comment by k4n30** [comment]
@mjmadsen & @Gobberwart where are we at with this?
